### PR TITLE
Hh/dtm 421/yml with variables

### PR DESCRIPTION
**Pull request by hal-henningsmoen**
Review ✅  Merge ✅ 
@dhallman this is ready to go.
This change adds yaml property back referencing employing the new ${} operator with or without the ^ qualifier. 
For a detailed description of this extension to basic YAML processing and example usage enter 
`hodor master:topic yaml_processing_extensions` 
at the bash command prompt

### CODE REVIEW

File: topics/master/yaml_processing_extensions.txt [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+The variables using back references would resolve to the following:
+:first_ggrandparent => "first",
+:first_grandparent => "o1",
+:first_parent => "p1",
+:first_child => "c1",
+:first_grandchild => "g1",
+:first_ggrandchild => "gg1"
+
+This is because when a reference is found the application walks back uo to the root of the currently defined tree of properties then
```
Comment (rdhallman) [inline topics/master/yaml_processing_extensions.txt]: @hal-henningsmoen Is this characterization accurate? I thought we decided not to "look" for the value by traversing the hierarchy looking for the "nearest" value in our recent HipChat conversation. I'm fine with either approach, just wanted to make sure the documentation matches the decision.

File: lib/hodor/environment.rb [hunk 2]
```diff
@@ -241,6 +242,48 @@ def ssh_addr
...
+        val.each_pair { |k, v|
+          exp_val = yml_expand(v, more_parents)
+          val[k] = exp_val
+        }
+      else
+        val
+      end
+    end
+
+    def yml_flatten(parent_key, val)
```
Comment (rdhallman) [inline lib/hodor/environment.rb:274]: Is yml_flatten being used anywhere yet? I had originally called it within one of the job.properties files.

File: topics/master/yaml_processing_extensions.txt [hunk 3]
```diff
@@ -0,0 +1,77 @@
+YAML Processing Extensions
+---------------------------------------------------------------------------------------------------
+
+Variable back-referencing:
+------------------------------
+
+In order to reduce repetition in .yml property files Hodor has implemented variable back referencing.
```
Comment (rdhallman) [inline topics/master/yaml_processing_extensions.txt]: I have a few sentences I'd like to contribute to this explanation. Will add that here shortly.

File: topics/master/yaml_processing_extensions.txt (hunk 1 above)
Comment (hal-henningsmoen) [inline topics/master/yaml_processing_extensions.txt]: Yes, this describes the functionality as implemented.  I believe the final decisions was to leave in the flexibility and specificity provided by using ^.

File: lib/hodor/environment.rb (hunk 2 above)
Comment (hal-henningsmoen) [inline lib/hodor/environment.rb:274]: I will need to verify.  Thanks for pointing this out.

Comment (hal-henningsmoen) [inline lib/hodor/environment.rb:274]: Will be moving this method so will leave as is for now.

File: topics/master/yaml_processing_extensions.txt [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    :name: production_data
+    :connection_string:  jdbc:mysql://${^msql_host}:${^mysql_port}/${name}
+
+Note that the connection string has three variable expansions, two that include '^' qualifiers. Both
+mysql_host and mysql_port are defined by the parent of connection_string, so the '^' qualifier causes
+Hodor to look to the parent for values. On the other hand, the ${name} variable is defined at the
+same hierarchical level, and therefore no '^' is required to locate its value.
+
+More precisely, the ^ qualifier is used to specify which of the referenced versions of the property to
+return if the property is defined at more than one level of the yml definitions.
```
Comment (rdhallman) [inline topics/master/yaml_processing_extensions.txt]: This doesn't quite sound right. The ^ qualifier looks up exactly one level in any condition, not just "if the property is defined at more than one level of the yml definition."  If the property is only defined at a single level, the ^ qualify will look up one level even with 1 definition for the property.

File: topics/master/yaml_processing_extensions.txt [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+:name: o1
+:family:
+  :name: p1
+    :children:
+      :name: c1
+      :grandchildren:
+        :name: g1
+        :ggrandchildren:
+          :name: gg1
+:ggrandparent: ${^^^^^name}
```
Comment (rdhallman) [inline topics/master/yaml_processing_extensions.txt]: The indention appears to be off here. With 0 indention :ggrandparent: would be a the root level of the hierarchy, so the ^^^^^ references wouldn't work at the root.

**Pull request merged**